### PR TITLE
[DB sources] : Add plumbing for adding transient errors

### PR DESCRIPTION
**Pull request by akashkulk**
Closes https://github.com/airbytehq/airbyte-internal-issues/issues/7107

Adds supports for transient errors and marks the transient trouble syncing from Postgres hot standby error as transient : https://docs.airbyte.com/integrations/sources/postgres/postgres-troubleshooting#sync-data-from-postgres-hot-standby-server

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/dependencies/build.gradle [hunk 1]
```diff
@@ -17,7 +17,7 @@ dependencies {
     api 'com.fasterxml.jackson.datatype:jackson-datatype-jsr310'
     api 'com.google.guava:guava:33.0.0-jre'
     api 'commons-io:commons-io:2.15.1'
-    api ('io.airbyte.airbyte-protocol:protocol-models:0.7.0') { exclude group: 'com.google.api-client', module: 'google-api-client' }
+    api ('io.airbyte.airbyte-protocol:protocol-models:0.9.0') { exclude group: 'com.google.api-client', module: 'google-api-client' }
```
Comment (evantahler) [inline airbyte-cdk/java/airbyte-cdk/dependencies/build.gradle:20]: I haven't looked at this closely before... why does out CDK have `com.google.api-client` as a sub-dependency?

Comment (akashkulk) [reply airbyte-cdk/java/airbyte-cdk/dependencies/build.gradle:20]: I'm also not sure. Is this something that the destination team needs for bigquery? cc : @stephane-airbyte

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt [hunk 2]
```diff
@@ -220,13 +220,22 @@ internal constructor(
             // exist, we
             // just return the original exception.
             ApmTraceUtils.addExceptionToTrace(e)
-            val rootThrowable = ConnectorExceptionUtil.getRootConfigError(e)
-            val displayMessage = ConnectorExceptionUtil.getDisplayMessage(rootThrowable)
+            val rootConfigErrorThrowable = ConnectorExceptionUtil.getRootConfigError(e)
+            val rootTransientErrorThrowable = ConnectorExceptionUtil.getRootTransientError(e)
+            // val displayMessage = ConnectorExceptionUtil.getDisplayMessage(rootThrowable)
```
Comment (rodireich) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt:225]: nit: fix formatting. clean comment

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt [hunk 3]
```diff
@@ -30,13 +31,18 @@ object ConnectorExceptionUtil {
...
     }
 
+    fun isTransientError(e: Throwable?): Boolean {
+        return isTransientErrorException(e) || isRecoveryConnectionException(e)
+    }
+
     fun getDisplayMessage(e: Throwable?): String? {
         return if (e is ConfigErrorException) {
             e.displayMessage
+        } else if (e is ConfigErrorException) {
```
Comment (rodireich) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt:44]: ??
Another ConfigErrorException

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/exceptions/TransientErrorException.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * An exception that indicates a transient error was encountered. This exception is caught and emits
+ * an AirbyteTraceMessage.
+ */
+class TransientErrorException : RuntimeException {
+    val displayMessage: String
+
+    constructor(displayMessage: String) : super(displayMessage) {
+        this.displayMessage = displayMessage
```
Comment (rodireich) [inline airbyte-cdk/java/airbyte-cdk/dependencies/src/main/kotlin/io/airbyte/commons/exceptions/TransientErrorException.kt:15]: You should be able to get it from super with `getMessage()`

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt (hunk 3 above)
Comment (akashkulk) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt:44]: oops. TransientErrorException

**Review by rodireich — APPROVED** [review]

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt [hunk 5]
```diff
@@ -30,13 +31,18 @@ object ConnectorExceptionUtil {
...
 
+    fun isTransientError(e: Throwable?): Boolean {
+        return isTransientErrorException(e) || isRecoveryConnectionException(e)
+    }
+
     fun getDisplayMessage(e: Throwable?): String? {
         return if (e is ConfigErrorException) {
             e.displayMessage
+        } else if (e is TransientErrorException) {
+            e.message
```
Comment (xiaohansong) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt:45]: shouldn't this be e.displayMessage?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt [hunk 6]
```diff
@@ -240,7 +250,11 @@ internal constructor(
                         .withConnectionStatus(
                             AirbyteConnectionStatus()
                                 .withStatus(AirbyteConnectionStatus.Status.FAILED)
-                                .withMessage(displayMessage)
+                                .withMessage(
+                                    ConnectorExceptionUtil.getDisplayMessage(
+                                        rootConfigErrorThrowable
```
Comment (xiaohansong) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt:255]: If errors are config error, and command is CHECK, we will not go into this special handling logic which might be a problem

Also is it true that errors in CHECK are always config errors?

**Review by xiaohansong — CHANGES_REQUESTED** [review]

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt (hunk 6 above)
Comment (akashkulk) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt:255]: Ah you're right. Good catch. Let me try to fix this

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt (hunk 5 above)
Comment (akashkulk) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionUtil.kt:45]: Ah no - we're using the parent method which is just message

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt (hunk 6 above)
Comment (akashkulk) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt:255]: just reordered this

Comment (xiaohansong) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt:242]: could `rootConfigErrorThrowable` be null here if check encountered a non config error?

Comment (akashkulk) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/base/IntegrationRunner.kt:242]: See `getRootConfigError`. It returns the top-level exception in that case

**Review by xiaohansong — APPROVED** [review]

**Comment by edgao** [comment]
just dropping in to say thanks for implementing this :rocket: one less yak on my to-shave list

**Pull request merged**